diesel/src/sqlite/connection/mod.rs

```diff
@@ -372,4 +372,105 @@ mod tests {
             .get_result::<(i32, i32, i32)>(&connection);
         assert_eq!(Ok((2, 3, 4)), added);
     }
+
+    use crate::sql_types::Foldable;
+    sql_function! {
+        #[aggregate]
+        fn my_sum<ST: Foldable>(expr: ST) -> ST::Sum;
+    }
+
+    struct MySum<T: Foldable> {
+        sum: T::Sum,
+    }
+    impl<T: Foldable> MySum<T> {
+        fn new() -> Self {
+            Self { sum: 0 }
+        }
+    }
+    impl<T: Foldable> Aggregator<T> for MySum<T> {
+        fn step(&mut self, expr: T) {
+            self.sum += expr;
+        }
+
+        fn finalize(&self) -> T::Sum {
+            self.sum
+        }
+    }
+    table! {
+        my_sum_example {
+            id -> Integer,
+            value -> Integer,
+        }
+    }
+
+    #[test]
+    fn register_aggregate_function() {
+        use self::my_sum_example::dsl::*;
+
+        let connection = SqliteConnection::establish(":memory:").unwrap();
+        connection.execute("CREATE TABLE my_sum_example (id integer primary key autoincrement, value integer)").unwrap();
+        connection.execute("INSERT INTO my_sum_example (value) VALUES (1), (2), (3)").unwrap();
+
+        let my_sum_struct = MySum::new();
+        my_sum::register_impl(&connection, my_sum_struct).unwrap();
+
+        let result = crate::select(my_sum(value)).get_result::<i32>(&connection).unwrap();
+        assert_eq!(6, result);
+    }
+
+    use crate::sql_types::SqlOrd;
+    sql_function! {
+        #[aggregate]
+        fn range_max<T: SqlOrd>(exprs: [T; 3]) -> T;
+    }
+
+    struct RangeMax<T: SqlOrd> {
+        max_value: Option<T>,
+    }
+    impl<T: SqlOrd> RangeMax<T> {
+        fn new() -> Self {
+            Self { max_value: None }
+        }
+    }
+    impl<T: SqlOrd> Aggregator<T> for RangeMax<T> {
+        fn step(&mut self, exprs: [T; 3]) {
```

Comment:
This variant would not allow the arguments to have different types, which is a quite hard assumption for the general case.
Using the variant of the `Aggregator` trait proposed above would allow that by using a tuple here. Unfortunately we then need to decompose the tuple as argument here, as we cannot relay on the compiler magic happening for the `Fn` traits.